torch/_library/define.py

```diff
@@ -0,0 +1,334 @@
+            >>>
+            >>> @custom_nonzero.impl_abstract
+            >>> def _(x):
+            >>>     # Number of nonzero-elements is data-dependent.
+            >>>     # Since we cannot peek at the data in an abstract impl,
+            >>>     # we use the ctx object to construct a new symint that
+            >>>     # represents the data-dependent size.
+            >>>     ctx = torch.library.get_ctx()
+            >>>     nnz = ctx.new_dynamic_size()
+            >>>     shape = [nnz, x.dim()]
+            >>>     result = x.new_empty(shape, dtype=torch.int64)
+            >>>     return result
+            >>>
+            >>> x = torch.tensor([0, 1, 2, 0, 0, 1])
+            >>> # xdoctest: +SKIP("Requires Python <= 3.11")
+            >>> out = torch.compile(custom_nonzero)(x)
+            >>> assert torch.allclose(out, x.nonzero())
+
+        """
+        self._abstract_fn = fn
+        return fn
+
+    def _build(self):
+        lib = self._lib
+        lib.define(f"{self._name}{self._schema}")
+        self._opoverload = _library.utils.lookup_op(self._qualname)
+
+        def abstract_impl(*args, **kwargs):
+            if self._abstract_fn is None:
+                import torch
+
+                raise torch._subclasses.fake_tensor.UnsupportedOperatorException(
+                    self._opoverload
+                )
+            return self._abstract_fn(*args, **kwargs)
+
+        library.impl_abstract(self._qualname, lib=lib)(abstract_impl)
+
+    def __call__(self, *args, **kwargs):
+        return self._opoverload(*args, **kwargs)
+
+
+OPDEF_TO_LIB: Dict[str, "library.Library"] = {}
+
+
+def get_library_allowing_overwrite(namespace: str, name: str) -> "library.Library":
+    qualname = f"{namespace}::{name}"
+
+    if qualname in OPDEF_TO_LIB:
+        OPDEF_TO_LIB[qualname]._destroy()
+        del OPDEF_TO_LIB[qualname]
+
+    lib = library.Library(namespace, "FRAGMENT")
+    OPDEF_TO_LIB[qualname] = lib
+    return lib
+
+
+def mangle_module(module):
+    """Mangles the module name.
+
+    The scheme is replacing dots with some number of underscores
+    (specified as mangledN where N is the number of underscores).
+
+    Examples:
+    foo.bar.baz -> mangled1_foo_bar_baz
+    foo_bar.baz -> mangled2__foo_bar__baz
+    foo.__baz__ -> mangled3___foo_____baz__
+
+    Don't parse the mangled string directly; use mangle_module and demangle_module
+    """
+    sep = unique_underscoring(module)
+    prefix = f"mangled{len(sep)}"
+    splits = module.split(".")
+    return sep.join([prefix, *splits])
+
+
+def unique_underscoring(s: str):
+    i = 1
+    while True:
+        result = "_" * i
+        if result not in s:
+            return result
+        i += 1
+
+
+def get_caller_module(stacklevel=1):
+    """Returns the fully qualified name of the module that called this function."""
+    frame = sys._getframe(stacklevel)
+    mod = inspect.getmodule(frame)
+    return mod
```

Comment:
How come we can't just read the module off of the function that the decorator is called with?

Reply:
We can! thanks for pointing it out.